packages/core-plugin-api/src/translation/types.ts

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright 2023 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+export interface TranslationRefConfig<
+  Messages extends Record<keyof Messages, string>,
+> {
+  id: string;
+  messages: Messages;
+  lazyResources?: Record<string, () => Promise<{ messages: Messages }>>;
+  resources?: Record<string, Messages>;
```

Comment:
I totally agree with you and we are on the way right? Something i want to discuss 

1. about `createTranslationResource`, why not `messages` define all kind language messages. like 

```
createTranslationResource({
  ref: userSettingsTranslationRef,
  messages: {
     en: () => import('./en/user-settings'),
     zh: () => import('./zh/user-settings'),
  }
})
``` 

2. code above 
```
createTranslationResource({
  ref: userSettingsTranslationRef,
  language: 'zh',
  messages: async () => createTranslationResource({
    ref: userSettingsTranslationRef,
    language: 'zh',
    messages: {
      select_lng: '选择中文-app',
    },
  })
})
```
why use `createTranslationResource` in `createTranslationResource` for messages loading? sorry, I don't get your mind

Reply:
For (1), the idea is that integrators would be able to select the specific languages that they would want to have available, rather than having it be all or nothing. I do anticipate that we'll be able to pair this up with some utilities that essentially load in all translations of a particular languages from all plugins and packages.

The (2) example is simplified, here's a more complete example:

```ts
// translations.ts
export const zh = createTranslationResource({
  ref: userSettingsTranslationRef,
  language: 'zh',
  messages: () => import('./zh'),
});

// translations/zh.ts
export default createTranslationResource({
  ref: userSettingsTranslationRef,
  language: 'zh',
  messages: {
    select_lng: '选择中文-app',
  },
});
```

The `createTranslationResource` in `translations.ts` serves to synchronously associate the messages with a particular translation ref and language, while `zh.ts` contains the actual translated messages, created with the help of `createTranslationResource` and the translation ref. These two layers are needed since we don't want to eagerly load in `zh.ts`, the entire purpose of the lazy loading is to avoid loading it until it is needed. But before we've loaded it we don't know what translations it contain, which is why we create the outer reference that tells the i18n system which namespace and language the resource is for.

Reply:
For 1, [rather than having it be all or nothing], you mean mostly to type safe, right. Because `Record<string, Messages>` can be no keys.

For 2, may need some time to refactor, At pulgin level 

```ts
export const userSettingsTranslationRef = createTranslationRef({
  id: 'user-settings',
  messages: {
    select_theme_custom: 'Select {{custom}}',
    lng: '{{lng}}',
    select_lng: 'Select language {{lng}}',
  },
  lazyResources: {
    en: () => import('./resource/en.ts'),
    zh: () => import('./resource/zh.ts')
  }
});
```
do your think `lazyResources` likewise similar to `createTranslationResource ` options 

```ts

 lazyResources: [
      {
      language: 'en',
      messages: () => import('./resource/en.ts')
    },
    {
      language: 'zh',
      messages: () => import('./resource/zh.ts')
    }
  ]
```